### PR TITLE
Implement some more of the WASI API

### PR DESCRIPTION
**Pull request by sbc100**
In order to test our WASI API implementation I've added
`wasi-test-suite` as submodule.  This submodule is only
needed for testing.

### CODE REVIEW

File: emscripten.py
Comment (kripken) [inline emscripten.py:83]: `building.py`'s `metadce()` has a `WASI_IMPORTS` list - maybe we can move that to a global position?

File: src/library_syscall.js
Comment (kripken) [inline src/library_syscall.js:1451]: this deserves a comment. I'm guessing this is because we need to check for growth on each call, without a proper notification hook?

File: tests/test_other.py
Comment (kripken) [inline tests/test_other.py:9647]: It's a little funny to set `PURE_WASI` and also NODERAWFS here... I guess this is the "normal" mode wasi works in, while emscripten's default is to sandbox the filesystem entirely. Anyhow, just a thought, I don't have an idea for an improvement.

File: src/library_wasi.js
Comment (kripken) [inline src/library_wasi.js:268]: what is 3, and why is it `/`?

File: src/library_wasi.js
Comment (kripken) [inline src/library_wasi.js:353]: maybe a TODO in this function?

File: src/library_wasi.js
Comment (kripken) [inline src/library_wasi.js:386]: technically these should be `rightsBase` etc. in `camelCase`. But maybe it makes sense to have an exception here?

File: tests/test_other.py
Comment (sbc100) [reply tests/test_other.py:9647]: I think it makes sense.   PUSE_WASI is a description of the way the wasm file interacts with JS, it describes the shape of the syscall layer.    noderawfs is effecively a plugin the JS-based filesystem code right?  It should work just fine whatever the shape of the syscall layer that goes from wasm to JS.

File: src/library_wasi.js
Comment (sbc100) [reply src/library_wasi.js:268]: Clearly this needs work yes :)

This basically says expose the whole of emscriptens VFS (the virtual filesytem created by the FS code) and make that available to WASI via the preopen system.   Pre-open file descriptors start at 3, right after stdin, stdout and stderr.   

This really needs integration with the FS layer... I think I should probably fix that even before landing this..

**Comment by caspervonb** [comment]
Is this blocked on anything? Anything I can do on the tests side? 🙃

**Comment by sbc100** [comment]
> Is this blocked on anything? Anything I can do on the tests side?

No, just lack time :)    I will try to get this landed before the end of the year.

**Comment by sbc100** [comment]
@kripken I'm trying to revive this old PR.  I've improved it a bunch and it runs more tests now.

File: src/library_syscall.js
Comment (kripken) [inline src/library_syscall.js:1000]:
```suggestion
  // In PURE_WASI_ mode we can't assume the wasm binary was built by emscripten
```

File: src/library_wasi.js
Comment (kripken) [inline src/library_wasi.js:83]: local debug?

File: src/shell.js
Comment (kripken) [inline src/shell.js:221]: Is this related to this PR?

**Review by kripken — APPROVED** [review]
lgtm % questions

File: src/shell.js
Comment (sbc100) [reply src/shell.js:221]: Yes, one of the tests here verifies the contents of arg0.   

This fixes a (what I claim is a) bug where were not honoring the existing `{{{ makeModuleReceive('thisProgram') }}}` under node.

**Comment by kripken** [comment]
The emscripten roller seems to be erroring after this landed. Seems like something with the new submodule directory?

https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket/8784982667442317537/+/u/Emscripten_testsuite__cores_/stdout

**Comment by sbc100** [comment]
Should be fixed in https://chromium-review.googlesource.com/c/emscripten-releases/+/4391454

**Pull request merged**
